File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/scheduling/OperationScheduleAssignment.kt
package stasis.client_android.lib.ops.scheduling

import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.schedules.ScheduleId
import java.nio.file.Path

sealed class OperationScheduleAssignment {
    abstract val schedule: ScheduleId

    data class Backup(
        override val schedule: ScheduleId,
        val definition: DatasetDefinitionId,
        val entities: List<Path>
    ) : OperationScheduleAssignment()

    data class Expiration(
        override val schedule: ScheduleId
    ) : OperationScheduleAssignment()

    data class Validation(
        override val schedule: ScheduleId
    ) : OperationScheduleAssignment()

    data class KeyRotation(
        override val schedule: ScheduleId
    ) : OperationScheduleAssignment()
}

File: client-ios/lib/Sources/StasisClientLib/Ops/Scheduling/OperationScheduleAssignment.swift
import Foundation

public enum OperationScheduleAssignment: Sendable, Equatable, Hashable {
    case backup(schedule: ScheduleId, definition: DatasetDefinitionId, entities: [URL])
    case expiration(schedule: ScheduleId)
    case validation(schedule: ScheduleId)
    case keyRotation(schedule: ScheduleId)

    public var schedule: ScheduleId {
        switch self {
        case .backup(let schedule, _, _): schedule
        case .expiration(let schedule): schedule
        case .validation(let schedule): schedule
        case .keyRotation(let schedule): schedule
        }
    }
}
